File: fren/app/src/main/java/com/gtek/fren/ui/home/HomeViewModel.java
package com.gtek.fren.ui.home;

import android.app.Application;
import android.util.Log;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.camera.core.ImageProxy;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.google.android.gms.tasks.Task;
import com.google.android.gms.tasks.Tasks;
import com.gtek.fren.ui.helper.EmotionBenchmark;
import com.gtek.fren.ui.helper.EmotionClassifier;
import com.gtek.fren.ui.helper.ImageProcessor;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class HomeViewModel extends AndroidViewModel {
    private static final String TAG = "HomeViewModel";

    private final MutableLiveData<List<EmotionClassifier.EmotionResult>> _emotionResults = new MutableLiveData<>();
    public LiveData<List<EmotionClassifier.EmotionResult>> emotionResults = _emotionResults;

    private final MutableLiveData<String> _processingError = new MutableLiveData<>();
    public LiveData<String> processingError = _processingError;

    private final MutableLiveData<Boolean> _isInitialized = new MutableLiveData<>(false);
    public LiveData<Boolean> isInitialized = _isInitialized;

    private final ExecutorService executorService = Executors.newFixedThreadPool(2);
    private ImageProcessor imageProcessor;
    private EmotionClassifier emotionClassifier;
    private final EmotionBenchmark benchmark;


    public static class HomeViewModelFactory implements ViewModelProvider.Factory {
        private final Application application;
        private final EmotionBenchmark benchmark;

        public HomeViewModelFactory(Application application) {
            this.application = application;
            this.benchmark = new EmotionBenchmark(); // Create benchmark instance here
        }

        @NonNull
        @Override
        public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
            if (modelClass.isAssignableFrom(HomeViewModel.class)) {
                return (T) new HomeViewModel(application, benchmark);
            }
            throw new IllegalArgumentException("Unknown ViewModel class");
        }
    }
    public HomeViewModel(@NonNull Application application, EmotionBenchmark benchmark) {
        super(application);
        this.benchmark = new EmotionBenchmark();
        initializeEmotionClassifier(application);
    }

    private void initializeEmotionClassifier(Application application) {
        executorService.execute(() -> {
            try {
                emotionClassifier = new EmotionClassifier(application.getApplicationContext());
                _isInitialized.postValue(true);
                Log.d(TAG, "EmotionClassifier initialized successfully");
            } catch (Exception e) {
                Log.e(TAG, "Failed to initialize EmotionClassifier: " + e.getMessage());
                _processingError.postValue("Failed to initialize classifier: " + e.getMessage());
                _isInitialized.postValue(false);
            }
        });
    }

    public void initializeImageProcessor(ImageView overlayView) {
        if (emotionClassifier != null && overlayView != null) {
            imageProcessor = new ImageProcessor(
                    executorService,
                    emotionClassifier,
                    _emotionResults,
                    _processingError,
                    overlayView,
                    benchmark
            );
            Log.d(TAG, "ImageProcessor initialized with overlay view");
        } else {
            Log.e(TAG, "Cannot initialize ImageProcessor: emotionClassifier or overlayView is null");
        }
    }

    public void startDetection() {
        if (imageProcessor != null) {
            imageProcessor.setIsDetecting(true);
            _emotionResults.postValue(new ArrayList<>());
            Log.d(TAG, "Detection started");
        } else {
            Log.e(TAG, "Cannot start detection: ImageProcessor is null");
        }
    }

    public void stopDetection() {
        if (imageProcessor != null) {
            imageProcessor.setIsDetecting(false);
            _emotionResults.postValue(new ArrayList<>());
            Log.d(TAG, "Detection stopped");
        }
    }

    public Task<Void> processImageWithFaceDetection(final ImageProxy imageProxy) {
        if (Boolean.FALSE.equals(_isInitialized.getValue())) {
            Log.e(TAG, "Cannot process image: System not yet initialized");
            return Tasks.forException(new IllegalStateException("System not initialized"));
        }

        if (imageProcessor != null) {
            return Tasks.call(() -> {
                imageProcessor.processImageWithFaceDetection(imageProxy);
                return null;
            });
        } else {
            Log.e(TAG, "Cannot process image: ImageProcessor is null");
            _processingError.postValue("System not properly initialized");
            return Tasks.forException(new IllegalStateException("ImageProcessor is null"));
        }
    }

    public void logPerformanceMetrics() {
        if (imageProcessor != null) {
            imageProcessor.logPerformanceMetrics();
        }
    }

    public void resetBenchmark() {
        if (imageProcessor != null) {
            imageProcessor.resetBenchmark();
        }
    }

    // Tambahkan method untuk mengecek status ImageProcessor
    public boolean isProcessorReady() {
        return imageProcessor != null;
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        cleanup();
        executorService.shutdown();
        Log.d(TAG, "ViewModel cleared");
    }

    public void cleanup() {
        stopDetection();
        if (imageProcessor != null) {
            imageProcessor.cleanup();
            imageProcessor = null;
        }
        if (emotionClassifier != null) {
            emotionClassifier = null;
        }
    }
}

File: fren/app/src/main/java/com/gtek/fren/ui/home/HomeViewModel.kt
package com.gtek.fren.ui.home

import android.app.Application
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.ImageFormat
import android.graphics.Rect
import android.graphics.YuvImage
import android.util.Log
import androidx.annotation.OptIn
import androidx.camera.core.ExperimentalGetImage
import androidx.camera.core.ImageProxy
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.face.FaceDetector
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.opencv.android.Utils
import org.opencv.core.Core
import org.opencv.core.Mat
import org.opencv.core.Size
import org.opencv.imgproc.Imgproc
import java.io.ByteArrayOutputStream

data class EmotionResult(
    val emotion: String,
    val confidence: Float
)

private fun ImageProxy.toBitmap(): Bitmap {
    val yBuffer = planes[0].buffer
    val uBuffer = planes[1].buffer
    val vBuffer = planes[2].buffer

    val ySize = yBuffer.remaining()
    val uSize = uBuffer.remaining()
    val vSize = vBuffer.remaining()

    val nv21 = ByteArray(ySize + uSize + vSize)

    // Pastikan buffer tidak null dan memiliki data yang cukup
    yBuffer?.get(nv21, 0, ySize)
    vBuffer?.get(nv21, ySize, vSize)
    uBuffer?.get(nv21, ySize + vSize, uSize)

    return try {
        val yuvImage = YuvImage(nv21, ImageFormat.NV21, width, height, null)
        ByteArrayOutputStream().use { outputStream ->
            yuvImage.compressToJpeg(Rect(0, 0, width, height), 100, outputStream)
            val imageBytes = outputStream.toByteArray()
            BitmapFactory.decodeByteArray(imageBytes, 0, imageBytes.size)
                ?: throw IllegalStateException("Failed to decode bitmap")
        }
    } catch (e: Exception) {
        Log.e("ImageProxy", "Error converting to bitmap: ${e.message}")
        throw e
    }
}

class HomeViewModel(application: Application) : AndroidViewModel(application) {
    private val tag = "HomeViewModel"
    private val _emotionResults = MutableLiveData<List<EmotionResult>>()
    val emotionResults: LiveData<List<EmotionResult>> = _emotionResults

    private val _processingError = MutableLiveData<String>()
    val processingError: LiveData<String> = _processingError

    private var isDetecting = false
    private val emotionClassifier: EmotionClassifier = EmotionClassifier(application.applicationContext)

    fun startDetection() {
        isDetecting = true
        Log.d(tag, "Detection started")
    }

    fun stopDetection() {
        isDetecting = false
        _emotionResults.postValue(emptyList())
        Log.d(tag, "Detection stopped")
    }

    @OptIn(ExperimentalGetImage::class)
    fun processImageWithFaceDetection(
        imageProxy: ImageProxy,
        faceDetector: FaceDetector
    ) {
        if (!isDetecting) {
            imageProxy.close()
            return
        }

        viewModelScope.launch(Dispatchers.Default) {
            try {
                val mediaImage = imageProxy.image
                if (mediaImage == null) {
                    Log.w(tag, "Skipping frame - no image available")
                    imageProxy.close()
                    return@launch
                }

                val image = InputImage.fromMediaImage(
                    mediaImage,
                    imageProxy.imageInfo.rotationDegrees
                )

                withContext(Dispatchers.Main) {
                    faceDetector.process(image)
                        .addOnSuccessListener { faces ->
                            if (faces.isEmpty()) {
                                Log.d(tag, "No faces detected")
                                _emotionResults.postValue(emptyList())
                                return@addOnSuccessListener
                            }

                            viewModelScope.launch(Dispatchers.Default) {
                                faces.forEach { face ->
                                    try {
                                        val bitmap = imageProxy.toBitmap()
                                        val mat = Mat()
                                        Utils.bitmapToMat(bitmap, mat)

                                        val boundingBox = face.boundingBox
                                        if (isValidBoundingBox(boundingBox, mat)) {
                                            val faceRegion = mat.submat(
                                                org.opencv.core.Rect(
                                                    boundingBox.left,
                                                    boundingBox.top,
                                                    boundingBox.width(),
                                                    boundingBox.height()
                                                )
                                            )

                                            val processedFace = preprocessFace(faceRegion)
                                            val emotions = emotionClassifier.classify(processedFace)
                                            val significantEmotions = emotions.filter { it.confidence > 0.3f }

                                            withContext(Dispatchers.Main) {
                                                _emotionResults.postValue(significantEmotions)
                                            }

                                            // Cleanup resources
                                            faceRegion.release()
                                            processedFace.release()
                                        } else {
                                            Log.w(tag, "Invalid face bounding box: $boundingBox")
                                        }
                                        mat.release()

                                    } catch (e: Exception) {
                                        Log.e(tag, "Error processing face: ${e.message}", e)
                                        withContext(Dispatchers.Main) {
                                            _processingError.postValue("Error processing face: ${e.message}")
                                        }
                                    }
                                }
                            }
                        }
                        .addOnFailureListener { e ->
                            Log.e(tag, "Face detection failed: ${e.message}", e)
                            _processingError.postValue("Face detection failed: ${e.message}")
                        }
                        .addOnCompleteListener {
                            imageProxy.close()
                        }
                }
            } catch (e: Exception) {
                Log.e(tag, "Image processing error: ${e.message}", e)
                withContext(Dispatchers.Main) {
                    _processingError.postValue("Image processing error: ${e.message}")
                }
                imageProxy.close()
            }
        }
    }

    private fun isValidBoundingBox(boundingBox: Rect, mat: Mat): Boolean {
        return boundingBox.left >= 0 &&
                boundingBox.top >= 0 &&
                boundingBox.right <= mat.cols() &&
                boundingBox.bottom <= mat.rows()
    }

    private fun preprocessFace(faceMat: Mat): Mat {
        return Mat().also { processedMat ->
            try {
                // Resize to model input size
                Imgproc.resize(faceMat, processedMat, Size(48.0, 48.0))

                // Convert to grayscale
                Imgproc.cvtColor(processedMat, processedMat, Imgproc.COLOR_BGR2GRAY)

                // Normalize pixel values
                Core.normalize(processedMat, processedMat, 0.0, 1.0, Core.NORM_MINMAX)
            } catch (e: Exception) {
                Log.e(tag, "Error preprocessing face: ${e.message}", e)
                processedMat.release()
                throw e
            }
        }
    }

    override fun onCleared() {
        super.onCleared()
        stopDetection()
        Log.d(tag, "ViewModel cleared")
    }
}